### DISCUSSION TITLE
a script to return the selected attribute value as string

### DISCUSSION THREAD

**Discussion by Bricklam** [freeplane/freeplane#1245] [Help]
now since the attributes with multi-lines cannot be edited inline, after this feature request:
https://github.com/freeplane/freeplane/issues/1241

I'm building a script that can input the multi-line text directly to the selected attribute (Value),
and another one that retrieve the value of the selected attribute (to the clipboard) as a plain text

but the main problem is: I was unable find a way to get the selected attribute value,
I know how to get selected nodes or any single node component, but the problem is: there is no clear way
to get the selected attribute, when selecting only one attribute part, among many attributes in the node, I couldn'tmake the script get the Value of the currently selected attribute at all.
I don't want to use filtering or (find) to get a specific attribute by name or value, I just want to operate on the selected attribute
(just like we do on selected nodes), to get or set its value text.
I just want the script to run on one node, and one attribute (the value part specifically)

after many attempts: this is the best script I could reach, unfortunately it doesn't return any text, but it can just tell if the selected attribute part is Name or Value. (printing the result on the status bar)
it includes many unneeded import statements, just in case:

**here is the script:**
```
import org.freeplane.api.Controller
import org.freeplane.api.ControllerRO
import org.freeplane.features.attribute.AttributeController
import org.freeplane.features.attribute.Attribute
import org.freeplane.features.attribute.AttributeRegistry
import org.freeplane.features.attribute.AttributeSelection
import org.freeplane.features.attribute.AttributeSelection.SelectedAttribute
import org.freeplane.features.attribute.AttributeSelection.SelectedAttribute.SelectedPart
import org.freeplane.features.attribute.NodeAttribute
import org.freeplane.features.map.NodeModel
import org.freeplane.view.swing.map.MapView
import org.freeplane.view.swing.map.NodeView
import org.freeplane.view.swing.map.attribute.AttributeSelectionChangeListener
import org.freeplane.view.swing.map.attribute.AttributeTable
import org.freeplane.view.swing.map.attribute.AttributeView


def selectedNode = c.selecteds

if (selectedNode != null) {
    def attributeSelection = AttributeController.getAttributeSelection()

    if (!attributeSelection.isEmpty()) {
        def selectedAttributes = attributeSelection.getSelectedAttributes()
        def selectedAttribute = selectedAttributes.first().getSelectedAttribute()

        def selectedPart = selectedAttributes.first().getSelectedPart()

        if (selectedPart == AttributeSelection.SelectedAttribute.SelectedPart.NAME) {

	       c.statusInfo = selectedPart.toString()
            
        } else if (selectedPart == AttributeSelection.SelectedAttribute.SelectedPart.VALUE) {
            c.statusInfo = selectedPart.toString()          
        }
    } 
} 
```

I found this line inside the code file: "org.freeplane.view.swing.map.attribute.AttributeTable"
I thought it might be relevant to my use case, but I failed to utilize it to achieve may goal: (getting the value text out of the selected attribute Value cell)
`final String attrName = getAttributeTableModel().getValueAt(row, 0).toString();`

Thanks

**Comment by dpolivaev** [comment]
In your scripts, you should not use any internal APIs because they can change at any time.
If you need anything not available in the scripting API it should be added there before you can use it.
Usually extending it is easy enough to be implemented soon.

As always I appreciate if you explain your use case and not only formulate your feature request.

**Reply by Bricklam** [reply]
thanks for your advice. I agree that using the internal API is not a good idea.

my point is to make interacting with attributes directly or via scripts easier and faster:
I use attributes heavily, I find myself in many situation wanting to get or set some attribute text quickly and directly, without worring about inputting or getting multi-line text. and I find that if there is a way to identify the currently selected attribute I can bypass the attiribute inline editor and to access their content directly.  the same way in working with nodes (I mean targetting only the selected nodes in scripting, instead of filtering them to get the target node)
the same idea for attributes would be powerful: for example it can enable the user to set or get its value text once it selected,
(to input the clipboard into the selected attribute value or vice versa (get the attirbute value to the clipboard))
and here are some advantages: (for accessing attributes data by selection instead of by name or index):
- the ability to save or restore multi-line text directly by scripts (without having to open the attribute inline editor (which no longer can edit multi-line text))
- adding more flexiblity & speed in editing attributes: (select and apply the script).
- multiple attributes can have the same name, making it difficult to get the right one by filtering, instead it can be selected easily

anyway, if you decided to add a new scripting API in this regard, I would suggest something like this:
Selected Node > SelectedAttribute (single/first one in selection) > setValue(String)
Selected Node > SelectedAttribute (single/first one in selection)  > getValue(String)

Thanks

**Reply by dpolivaev** [reply]
I am generally available only on Friday-Sunday and therefore I am reading current messages carefully during the week. Therefore I can only share a general remark: if you have already implemented what you need using the internal API, export is using some "Proxy" class from the scripting plugin. I know that the choice of the names and separation in RO and writing interfaces is at least questionable, but because of the compatibility considerations, we stick to it. And I think you can find out where to contribute your code.

**Comment by EdoFro** [comment]
@Bricklam,

What about using a dialog to select the source/target attribute?

Here is a script that shows you a dialog with a button for each attribute, and when clicking one it copies its text to the clipboard.

```groovy
import groovy.swing.SwingBuilder
import java.awt.*

swingBuilder = new SwingBuilder()

def attrNames = node.attributes.names

def attrButton(attr){
    swingBuilder.button(
        text : attr, 
        actionPerformed : {
                def valor = node.attributes.getFirst(attr)
                textUtils.copyToClipboard(valor.toString())
            }
        )
}


def buttonPanel = {swingBuilder.panel(
        layout : new GridLayout(0,1)
    ){
        attrNames.each{a ->
            attrButton(a)
        }
    }
}


def dialogo = swingBuilder.dialog(
                title              : 'Copy attribute as string to clipboard',
                name               : 'attrTestDialog',
                modal              : false,
                locationRelativeTo : ui.frame,
                minimumSize        : new Dimension(30,70),
                layout             : new GridLayout(0,1),
                owner              : ui.frame,
                show               : true,
                pack               : true
            ) {
                buttonPanel()
            }
            

return "ready"

```

**Reply by Bricklam** [reply]
Thanks @EdoFro for this amazing Script. it works fine. and I liked the idea of the new dialog
it can do half of the task (getting the value of selected attribute, still the difficult part is setting the value for the selected attribute).

I was working on a script (two scripts) to allow getting or setting the value of the selected attribute (from the clipboard or to it).
 and finally I was able to get it work (it was a nasty workaround), since I wasn't able to use the internal API and it is also a bad idea.
 but it worked somehow and I will post the scripts once they are finished.

anyway, I'm sure that there is no way available in the Scripting API to get the selected attribute name or value, you can get it by index, or by name but not by selection). that is why I had to think another way.

**Reply by EdoFro** [reply]
@Bricklam
ok. here are the two scripts. Just add some hotkeys to them and you're done, no?

### Copy attribute's text to clipboard
```groovy
//Copy attribute's text to clipboard

import groovy.swing.SwingBuilder
import java.awt.GridLayout
import java.awt.Dimension

def attrNames = node.attributes.names

dialogo = new SwingBuilder().dialog(
                title              : 'Copy attribute as string to clipboard',
                name               : 'attrTestDialog',
                modal              : false,
                locationRelativeTo : ui.frame,
                minimumSize        : new Dimension(30,70),
                layout             : new GridLayout(0,1),
                owner              : ui.frame,
                show               : true,
                pack               : true
            ) {
                def buttonPanel = panel(
                        layout : new GridLayout(0,1)
                    ){
                        attrNames.each{attr ->
                            button(
                                text : attr, 
                                actionPerformed : {
                                    def valor = node.attributes.getFirst(attr)
                                    textUtils.copyToClipboard(valor.toString())
                                    dialogo.dispose()
                                }
                            )
                        }
                    }
                
            }
            

return "copy"


```


### Paste clipboard's text to attribute
```groovy
//paste clipboard's text to attribute

import groovy.swing.SwingBuilder
import java.awt.GridLayout
import java.awt.Dimension
import java.awt.Toolkit
import java.awt.datatransfer.DataFlavor



def attrNames = node.attributes.names

dialogo = new SwingBuilder().dialog(
                title              : 'Copy attribute as string to clipboard',
                name               : 'attrTestDialog',
                modal              : false,
                locationRelativeTo : ui.frame,
                minimumSize        : new Dimension(30,70),
                layout             : new GridLayout(0,1),
                owner              : ui.frame,
                show               : true,
                pack               : true
            ) {
                def buttonPanel = panel(
                        layout : new GridLayout(0,1)
                    ){
                        attrNames.each{attr ->
                            button(
                                text : attr, 
                                actionPerformed : {
                                    def valor = readClipboard()
                                    node.attributes.set(attr, valor)
                                    dialogo.dispose()
                                }
                            )
                        }
                    }
                
            }

return "paste"

def readClipboard() {
    return (String) Toolkit.getDefaultToolkit().getSystemClipboard().getData(DataFlavor.stringFlavor)
}


```

Hope this helps,

edo

**Comment by macmarrum** [comment]
@Bricklam Scripting-API docs can be accessed from within Freeplane: `Help->Freeplane API…` or online: https://docs.freeplane.org/api/

To access attribute value in Groovy, you can do
```groovy
def value = node[attribute_name]
```
For more details, please see https://docs.freeplane.org/api/org/freeplane/api/NodeRO.html#getAt(java.lang.String)

**Reply by Bricklam** [reply]
yes. I checked the Scripting-API, and I know about the available ways to access the attribute (by name, by index,..)
but my point was: to be able to operate directly on the selected attribute (the attribute with highlighted name or value in the attributes table) the same way we execute scripts on selected nodes (without having to filter them by name or index),
you can call it (access by selection). and I didn't find a way in the API to identify/get the selected attribute. or any of its parts.

**Reply by macmarrum** [reply]
Agreed. The Scripting API is missing a way to tell which attribute is selected.

I might have misunderstood your point.

Now that you presented your idea, the need to tell which attribute is selected is clear. Therefore it would be great if the Scripting API could be extended. Your investigations might be the key.
Or maybe you'll find the time and energy to contribute such an extension :smile:

**Reply by dpolivaev** [reply]
While using internal API quite often is a bad idea, submitting a pull request extending the scripting API might be an excellent move.
